src/cudnn/batchnorm.jl

```diff
@@ -68,57 +84,82 @@ function cudnnBNForward!(y::DenseCuArray{T}, g::DenseCuArray{T}, b::DenseCuArray
                           dx::DenseCuArray{T}, x::DenseCuArray{T}, dy::DenseCuArray{T},
-                          running_mean::DenseCuArray{T}, running_var::DenseCuArray{T},
+                          running_mean, running_var,
                           momentum; cache = nothing, eps = T(1e-5),
                           alpha = T(1), beta = T(0),
-                          dalpha = T(1), dbeta = T(0), training = true) where T<:Union{Float32, Float64}
-  if training
-    xd = cudnnTensorDescriptor(x)
-    dyd = cudnnTensorDescriptor(dy)
-    dxd = cudnnTensorDescriptor(dx)
-    gd = cudnnTensorDescriptor(CUDNN_TENSOR_NCHW, cudnnDataType(T), Cint(length(_wsize(x))), dim4(_wsize(x),Val(CUDNN_TENSOR_NCHW)))
-    if cache !== nothing
-      mean, ivar = cache.mean, cache.ivar
-      info("mean and ivar are fetched from the cache")
-    else
-      mean, ivar = CU_NULL, CU_NULL
-    end
+                          dalpha = T(1), dbeta = T(0), training = true, 
+                          track_stats = true) where T<:Union{Float32, Float64}
+  
+  if !track_stats
+    running_mean = CU_NULL
+    running_var = CU_NULL
+  end
 
-    if eps < CUDNN_BN_MIN_EPSILON
-      eps = CUDNN_BN_MIN_EPSILON
-    end
+  xd = cudnnTensorDescriptor(x)
+  dyd = cudnnTensorDescriptor(dy)
+  dxd = cudnnTensorDescriptor(dx)
+  gd = cudnnTensorDescriptor(CUDNN_TENSOR_NCHW, cudnnDataType(T), Cint(length(_wsize(x))), dim4(_wsize(x),Val(CUDNN_TENSOR_NCHW)))
+  if cache !== nothing
+    mean, ivar = cache.mean, cache.ivar
+    # info("mean and ivar are fetched from the cache")
+  else
+    mean, ivar = CU_NULL, CU_NULL
+  end
 
-    cudnnBatchNormalizationBackward(handle(), CUDNN_BATCHNORM_SPATIAL, scalingParameter(T, alpha), scalingParameter(T, beta), scalingParameter(T, dalpha), scalingParameter(T, dbeta), xd, x, dyd, dy, dxd, dx, gd, g, dg, db, eps, mean, ivar)
+  if eps < CUDNN_BN_MIN_EPSILON
+    eps = CUDNN_BN_MIN_EPSILON
+  end
+
+  if training
+    cudnnBatchNormalizationBackward(handle(), CUDNN_BATCHNORM_SPATIAL, 
+          scalingParameter(T, alpha), scalingParameter(T, beta), scalingParameter(T, dalpha), scalingParameter(T, dbeta), 
+          xd, x, dyd, dy, dxd, dx, gd, g, dg, db, eps, 
+          mean, ivar)
   else
-    ivar = 1 ./ sqrt.(reshape(running_var, _wsize(x)) .+ eps)
-    dx .= dy .* reshape(g, _wsize(x)) .* ivar
-    rdims = ((1:ndims(x)-2)..., ndims(x))
-    dg .= vec(sum(dy .* (x .- reshape(running_mean, _wsize(x))) .* ivar, dims=rdims))
-    db .= vec(sum(dy, dims=rdims))
+    cudnnBatchNormalizationBackward(handle(), CUDNN_BATCHNORM_SPATIAL, 
+          scalingParameter(T, alpha), scalingParameter(T, beta), scalingParameter(T, dalpha), scalingParameter(T, dbeta), 
+          xd, x, dyd, dy, dxd, dx, gd, g, dg, db, eps, 
+          running_mean, running_var)  
   end
 end
-  
+
+function rrule(::typeof(batchnorm), g, b, x, running_mean, running_var, momentum; kws...)
```

Comment:
```suggestion
function rrule(::typeof(batchnorm), g, b, x::DenseCuArray{T}, running_mean, running_var, momentum; kws...) where T<:CUDNNFloat
```
This should allow us to add a generic rrule in NNlib later.

Reply:
Ideally let's use AbstractGPUArray

Reply:
Removed in favour of adding something to NNlib first. This isn't required anyhow because Flux already has an adjoint.